src/WindowStateSaver.vala

```diff
@@ -0,0 +1,163 @@
+/*
+ * Copyright 2023 elementary, Inc. <https://elementary.io>
+ * SPDX-License-Identifier: GPL-3.0-or-later
+ */
+
+public class Gala.WindowStateSaver : GLib.Object {
+    private static unowned WindowTracker window_tracker;
+    private static GLib.HashTable<string, GLib.Array<Meta.Window?>> app_windows;
+    private static Sqlite.Database db;
+
+    public static void init (WindowTracker window_tracker) {
+        WindowStateSaver.window_tracker = window_tracker;
+        app_windows = new GLib.HashTable<string, GLib.Array<Meta.Window?>> (GLib.str_hash, GLib.str_equal);
+
+        var path = Path.build_filename (Environment.get_home_dir (), ".local", "share", "io.elementary.gala-windowstate.db");
+        var rc = Sqlite.Database.open_v2 (path, out db);
+
+        if (rc != Sqlite.OK) {
+            critical ("Cannot open database: %d, %s", rc, db.errmsg ());
+            return;
+        }
+
+        Sqlite.Statement stmt;
+        rc = db.prepare_v2 (
+            """
+            CREATE TABLE IF NOT EXISTS apps (
+                id           INTEGER PRIMARY KEY,
+                app_id       TEXT,
+                window_index INTEGER,
+                last_x       INTEGER,
+                last_y       INTEGER,
+                last_width   INTEGER,
+                last_height  INTEGER
+            );
+            """,
+            -1, out stmt
+        );
+
+        if (rc == Sqlite.OK) {
+            rc = stmt.step ();
+            if (rc == Sqlite.DONE) {
+                // disable synchronized commits for performance reasons
+                rc = db.exec ("PRAGMA synchronous=OFF");
+                if (rc != Sqlite.OK) {
+                    warning ("Unable to disable synchronous mode %d, %s", rc, db.errmsg ());
+                }
+
+                return;
+            }
+        }
+
+        critical ("Cannot create table 'apps': %d, %s", rc, db.errmsg ());
+    }
+
+    public static void on_map (Meta.Window window) {
+        var app_id = window_tracker.get_app_for_window (window).id;
+
+        if (app_id.has_prefix ("window:")) {
+            // if window failed to be identified, don't remember it
+            return;
+        }
+
+        if (window.window_type != Meta.WindowType.NORMAL) {
+            return;
+        }
+
+        if (!(app_id in app_windows)) {
+            app_windows[app_id] = new GLib.Array<Meta.Window?> ();
+        }
+
+        var window_index = find_window_index (window, app_id);
+        app_windows[app_id].insert_val (window_index, window);
+
+        var tracking_window = false;
+        db.exec (
+            "SELECT last_x, last_y, last_width, last_height FROM apps WHERE app_id = '%s' AND window_index = '%d';".printf (app_id, window_index),
```

Comment:
This does not seem to escape `app_id`. I guess app_id will never contain ' characters, but is there an easy way to escape it? (Same for the `INSERT` statement below.)